packages/nextjs/src/app-router/server/ClerkProvider.tsx

```diff
@@ -27,21 +26,35 @@ export async function ClerkProvider(
   props: Without<NextClerkProviderProps, '__unstable_invokeMiddlewareOnAuthStateChange'>,
 ) {
   const { children, dynamic, ...rest } = props;
-  let statePromise: Promise<null | AuthObject> = Promise.resolve(null);
-  let nonce = Promise.resolve('');
 
-  if (dynamic) {
-    if (isNext13) {
-      /**
-       * For some reason, Next 13 requires that functions which call `headers()` are awaited where they are invoked.
-       * Without the await here, Next will throw a DynamicServerError during build.
-       */
-      statePromise = Promise.resolve(await getDynamicClerkState());
-      nonce = Promise.resolve(await getNonceFromCSPHeader());
-    } else {
-      statePromise = getDynamicClerkState();
-      nonce = getNonceFromCSPHeader();
+  async function generateStatePromise() {
+    if (dynamic) {
+      if (isNext13) {
+        /**
+         * For some reason, Next 13 requires that functions which call `headers()` are awaited where they are invoked.
+         * Without the await here, Next will throw a DynamicServerError during build.
+         */
+        return Promise.resolve(await getDynamicClerkState());
+      } else {
+        return getDynamicClerkState();
+      }
+    }
+    return Promise.resolve(null);
+  }
+
+  async function generateNonce() {
+    if (dynamic) {
+      if (isNext13) {
+        /**
+         * For some reason, Next 13 requires that functions which call `headers()` are awaited where they are invoked.
+         * Without the await here, Next will throw a DynamicServerError during build.
+         */
+        return Promise.resolve(await getNonceFromCSPHeader());
+      } else {
+        return getNonceFromCSPHeader();
+      }
     }
+    return Promise.resolve('');
```

Comment:
[nit] convert to a guard clause

```suggestion
    if (dynamic) {
	    return Promise.resolve('');
    }
    if (isNext13) {
      /**
       * For some reason, Next 13 requires that functions which call `headers()` are awaited where they are invoked.
       * Without the await here, Next will throw a DynamicServerError during build.
       */
      return Promise.resolve(await getNonceFromCSPHeader());
    } else {
      return getNonceFromCSPHeader();
    }
```